src/pip/_internal/vcs/git.py

```diff
@@ -322,7 +335,39 @@ def get_remote_url(cls, location):
                 found_remote = remote
                 break
         url = found_remote.split(' ')[1]
-        return url.strip()
+        return cls._git_remote_to_pip_url(url.strip())
+
+    @staticmethod
+    def _git_remote_to_pip_url(url):
+        # type: (str) -> str
+        """
+        Convert a remote url from what git uses to what pip accepts.
+
+        There are 3 legal forms **url** may take:
+
+            1. A fully qualified url: ssh://git@example.com/foo/bar.git
+            2. A local project.git folder: /path/to/bare/repository.git
+            3. SPC shorthand for form 1: git@example.com:foo/bar.git
+
+        Form 1 is output as-is. Form 2 must be converted to URI and form 3 must
+        be converted to form 1.
+
+        See the corresponding test test_git_remote_url_to_pip() for examples of
+        sample inputs/outputs.
+        """
+        if re.match(r"\w+://", url):
+            # This must already valid. Pass it though as-is.
+            return url
+        if os.path.exists(url):
+            # A local bare remote (git clone --mirror).
+            # Needs a file:// prefix.
+            return pathlib.PurePath(url).as_uri()
```

Comment:
Does this work well if `url` is a Windows-style path (using backslashes)?

Reply:
Yes, the backslashes are converted to forward slashes:
```python
>>> import pathlib
>>> pathlib.PureWindowsPath(r"C:\foo\bar").as_uri()
'file:///C:/foo/bar'
```
The colon in the middle looks weird but [according to wikipedia](https://en.wikipedia.org/wiki/File_URI_scheme#Windows) is correct.

Reply:
Hmm, I see there are 5 different definitions of a `path_to_ur()` function. Would you rather I used one of those instead?

Reply:
Nah just use pathlib. Those functions are leftovers from Python 2 and we want to eventually get rid of them.

(For others: `PurePath.__new__()` creates a `PureWindowsPath` or `PurePosixPath` depending on the current platform, so this `PurePath(path).as_uri()` call automatically does the right thing everywhere. This was the part I didn’t get previously.)